### ISSUE TITLE
Parcel (or any other bundler) support

### ISSUE THREAD

**Issue by alanjds** [rafi16jan/micropython-wasm#1]
Trying to use micropython.js following the npmjs readmem somehow an internal `node_modules/micropython/index.js:113:49` could not be imported.
```js
// index.js
import mp_js from 'micropython';

/*
(async () => {
    await mp_js;
    mp_js.init(64 * 1024);
    mp_js.do_str("print('hello world')\n");
})();
*/
```

```js
🚨  /Users/alanjds/src/git/parcel-plugin-micropython/node_modules/micropython/index.js:113:49: Unexpected token, expected ";" (113:49)
  111 |     return await methods.do_str(pyjs);
  112 |   }
> 113 |   methods.register_module = async (module, code) {
      |                                                 ^
  114 |     if (!pyjs_initiated) return pyjs += '\n' + await methods.do_str(code, module, false);
  115 |     return await methods.do_str(code, module, true);
  116 |   }
```

May be possible to this package to have some specification issue on `packages.json`?
Looks like some babel translation is not being applied? But I could not find it by myself. Am I looking at the wrong place/path?

Context:

```bash
$ node --version
v9.11.2
```

```json
$ cat package.json
{
  "name": "helloworld",
  "version": "1.0.0",
  "description": "An absolute bare-bones web app.",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "parcel serve web/index.html"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "eslint": "^6.0.1",
    "parcel-plugin-micropython": "file:../.."
  },
  "devDependencies": {
    "@babel/core": "^7.2.2"
  }
}
```

**Comment by rafi16jan** [comment]
This is babel on browser correct? I have only test this module with create-react-app template. And it should have work with other webpack boiler plates too. Should I add something to support plain babel or stuff?

**Comment by rafi16jan** [comment]
Ohhh sorry, it actually is a syntax error, should be 'async function' or 'async () =>'

**Comment by rafi16jan** [comment]
@alanjds I made a fix, please update your version. Thanks for reporting

**Comment by alanjds** [comment]
Thank you for the bugfix and release.

Now something else came up:

```js
🚨  /Users/alanjds/src/git/parcel-plugin-micropython/node_modules/micropython/!raw-loader!./js.py: ENOENT: no such file or directory, open '/Users/alanjds/src/git/parcel-plugin-micropython/node_modules/micropython/!raw-loader!./js.py'
Error: ENOENT: no such file or directory, open '/Users/alanjds/src/git/parcel-plugin-micropython/node_modules/micropython/!raw-loader!./js.py'

⚠️  /Users/alanjds/src/git/parcel-plugin-micropython/node_modules/micropython/lib/micropython.js:55:42: Cannot statically evaluate fs argument
  53 |               }
  54 |           } else if (process.argv[i].includes('.py')) {
> 55 |               contents += fs.readFileSync(process.argv[i], 'utf8');
     |                                          ^
  56 |               repl = false;;
  57 |           }
  58 |       }
```

Is this another bug or am I missing something?

**Comment by rafi16jan** [comment]
@alanjds that is weird, your Babel (or any transpiler that you use) doesn't evaluate !raw-loader! as a webpack loader https://github.com/webpack-contrib/raw-loader. Maybe webpack is **mandatory** for now. But I must clarify, you are not running it on Node.js didn't you? Because if you are, don't transpile it. Just directly require() the micropython module and you are all set.

**Comment by rafi16jan** [comment]
Can I fork this parcel-plugin-micropython repo? I want to test it myself and probably creating a PR for you if I found a workaround

**Comment by alanjds** [comment]
> Can I fork this parcel-plugin-micropython repo?

Sure.

The repo is yet to be pushed. As soon as got pushed to Github I will link it here.

> But I must clarify, you are not running it on Node.js didn't you?

No. I am trying to add browser support for Python

**Comment by alanjds** [comment]
@rafi16jan I invited you to the private repo on https://github.com/alanjds/parcel-plugin-micropython/

As soon as it got minimally usable I will open it to the world as Apache.

**Comment by alanjds** [comment]
The far I got researching Parcel issue tracker is this issue (https://github.com/parcel-bundler/parcel/pull/1101) to enable babel on symlinked modules, but is already merged. So, no path from here.

Thanks for the hand here.

**Comment by alanjds** [comment]
Today I tried to remove the `!raw-loader!` part and Parcel imported it as raw text, as it already have built-in the feature of import unknown formats as text.

Had to do the same with the `arraymemory!` on micropython.js somewhere I guess.

Will update you later.

**Comment by rafi16jan** [comment]
@alanjds sorry for not updating, I have been busy this week working on IoT projects. The other loader, is an ArrayBuffer loader. Required to read micropython.binary as an ArrayBuffer object. And the furthest that I can find about parcel bundler support for ArrayBuffer is this [issue](https://github.com/parcel-bundler/parcel/issues/2504).

Btw, I can't find your changes that are relevant to micropython in the private repo. I concur that it is a fork from dart loader plugin, but it looks like I will have some time figuring out how do I replace dart compilation to micropython :D

**Comment by rafi16jan** [comment]
I found that parcel have fs support. Maybe this bit of code that should only work on node.js
```javascript
  const file = require('fs').readFileSync(__dirname + '/micropython.binary');
  Module.emterpreterFile = file.buffer.slice(file.byteOffset, file.byteOffset + file.byteLength);
```
Could work on parcel since they have support for fs module.

Can you help me set up my micropython module on the private repo? Because I doesn't have time for now to figure it out myself. Thank you

**Comment by rafi16jan** [comment]
I made a new change at 6c289b3. Basically it changes the Node.js detection to Webpack detection. Previously it detects if this script is not running on the browser use fs module. If this script runs on the browser use webpack loaders. Now I change it to, if the script is running on webpack it use the webpack loaders. And if it is not, use fs module. This should support any bundler that have fs module, _unless_ the bundler will still evaluate !loader-name! requires. In this case, try to remove the line.

**Comment by alanjds** [comment]
Thanks. I will try your changes later.

> Btw, I can't find your changes that are relevant to micropython in the private repo.

I am still trying to `import micropython`. As it worked, next change will be to implement `doCompile`, creating the js needed. I am still trying to understand if will upload the verbatim `.py` file to be input on micropython or if will put on some storage accessible by micropython to be `import`-ed

**Comment by alanjds** [comment]
> This should support any bundler that have fs module, unless the bundler will still evaluate !loader-name! requires. In this case, try to remove the line.

Yes. This makes sense. At least for my limited knowledge of how JS bundlers work.

**Comment by rafi16jan** [comment]
@alanjds sorry, I forgot to publish the changes to NPM. Please check for update

**Comment by alanjds** [comment]
> Btw, I can't find your changes that are relevant to micropython in the private repo. I concur that it is a fork from dart loader plugin, but it looks like I will have some time figuring out how do I replace dart compilation to micropython :D

The relevant changes are on this folder: https://github.com/alanjds/parcel-plugin-micropython/tree/poc-mpy/src

But are yet far from working.

Right now I am still trying to make the basic wasm to be working nice on front-end as a dependency. Then I will pursue the task of allowing the `.py` files to be loaded by the running micropython.js. Maybe precompile to `.mpy`, but this is another whole history

**Comment by alanjds** [comment]
> @alanjds sorry, I forgot to publish the changes to NPM. Please check for update

I saw your changes and they make sense. However parceljs is not smart enough to ignore the webpack-specific imports, even surrounded by `if`s. I sent you a PR https://github.com/rafi16jan/micropython-wasm/pull/2 to fix it.

Does not solves the world, but at least compiles :/

**Comment by rafi16jan** [comment]
> > @alanjds sorry, I forgot to publish the changes to NPM. Please check for update
> 
> I saw your changes and they make sense. However parceljs is not smart enough to ignore the webpack-specific imports, even surrounded by `if`s. I sent you a PR #2 to fix it.
> 
> Does not solves the world, but at least compiles :/

Thank you, please answer my response on the PR.
